### DISCUSSION TITLE
Add `deal` support in `testActions`

### DISCUSSION THREAD

**Discussion by Rubilmax** [wevm/viem#2819] [Idea / Feature Request] [answered]
Include the `deal` utility to viem's built-in `testActions` in order to ease viem-based test developments for `setStorageAt`-compatible networks

`deal` is heavily inspired by foundry's [`deal` cheatcode](https://book.getfoundry.sh/cheatcodes/deal?highlight=deal#deal): it allows to deal any amount of most ERC20 tokens to any account on the test network, which avoids having to impersonate whales (because whale A may not be a whale anymore when the fork's block number is changed later in time - and it happens a lot throughout a software lifetime)

> [!IMPORTANT]  
> I am the developer & maintainer of [`viem-deal`](https://github.com/Rubilmax/viem-deal)

**Comment by jxom** [comment]
Looks great! I will accept a PR for this.

**Reply by Rubilmax** [reply]
Great, I'll work on that asap.

Did you get a chance to look at how I handle caching, to avoid brute-forcing the balanceOf storage slot every run?

Is there some other way you'd like caching to be handled in viem, or is a `./cache/deal.json` cache file enough for you?

I'm also thinking about moving it to `~/.foundry/cache/`, because it is included in foundry's GitHub action which avoids having to handle that cache file differently

**Reply by jxom** [reply]
We can just use the existing LRU cache: https://github.com/wevm/viem/blob/main/src/utils/lru.ts

**Reply by Rubilmax** [reply]
Right, so that answers how I should handle the cache in memory, but where do I store it in the file system (if available in the runtime environment)?

**Reply by jxom** [reply]
What is the perf impact of not having it? Wouldn’t Anvil be nearly instantaneous for storage slot access over a local network?

**Reply by Rubilmax** [reply]
It depends on the configuration of the runtime environment because anvil caches RPC queries

So having RPC cache + storage slot cache makes it instantaneous

Having RPC cache but no storage slot cache takes 20ms for most tokens and up to 100ms for complex tokens where the storage slot is at index 50 (e.g. aave tokens)

Having no RPC cache but storage slot cache takes the time of an RPC call and it's ok imo because the environment has no RPC cache anyway

Having no RPC cache and no storage slot cache is obv the slowest, and it's ok imo because it only happens the first time

Case 2 would be the most common case if there is no storage slot cache, so the answer to your question is 20-100ms as measured with vitest. For tests where you deal multiple times, I do think it's worth having a storage slot cache. What's the downside of storing it at the same place as foundry's cache?

**Reply by jxom** [reply]
> the answer to your question is 20-100ms as measured with vitest

This seems like an acceptable startup time with no warmed cache. 

I am open to introducing an abstraction for FS caching, but if the optimization is just to cut 20-100ms from initial setup, then I don't think it's really worth it? IMO, the in-memory cache (for storage slot access post-setup) should suffice for this use-case.

**Reply by roninjin10** [reply]
@Rubilmax I looked at your implementation out of curiousity. Is the slowness you are worried about mostly because it's finding the storage slot via trial and error?

If so what do you think about finding the storage slot instead via requesting an access list while doing a balanceOf call? That should get you the storage slot in a single request.

**Reply by Rubilmax** [reply]
Right, I didn't think about an in-memory cache so I was thinking about 20-100ms at every deal. I agree it's ok to have this delay at the first only, using an in-memory cache such as the LRU map you suggested @jxom 

Very interesting idea @roninjin10 , I'll add this to the issue on foundry!

**Comment by roninjin10** [comment]
+1 to adding this would like this to get inherited by [tevm](https://github.com/evmts/tevm-monorepo)

**Comment by jxom** [comment, marked as answer]
Maybe this is worth upstreaming in Anvil itself too: https://github.com/foundry-rs/foundry/issues/8956

**Reply by roninjin10** [reply]
If it ends up in anvil instead I will still happily take the js action in Tevm. Let me know if you want contribution credit for it in that case otherwise I’ll just copy paste the implementation

**Reply by Rubilmax** [reply]
Best would indeed to have it included in `anvil`. I'll push for this
